### PR TITLE
Introduce PHPStan

### PR DESCRIPTION
**Pull request by szepeviktor**
Closes #1986

### CODE REVIEW

**Comment by szepeviktor** [comment]
@jarednova I suggest you to start with `vendor/bin/phpstan analyze -l 0` and raise Level.

- 2: poorly written code
- 4: normal
- 8: robust!

**Comment by szepeviktor** [comment]
PHP 7.4 does not like this

```
1) TestTimberPost::testPostWithVideo

PHPUnit\Framework\Exception: Argument #2 (No Value) of PHPUnit\Framework\Assert::assertRegExp() must be a string

/home/travis/build/timber/timber/tests/test-timber-post.php:937
```

**Comment by gchtr** [comment]
I like this, a lot! 🤩

I think this could also help with https://github.com/timber/timber/issues/1863. And I would even fix issues up to a certain level *before* we apply the WordPress Coding Standards to the code base as one of the last steps for 2.0.

When I run it locally, it’s quite fast, so we could eventually think about adding this to Travis.

@jarednova Do you know why there are so many errors in Travis?

For now, I would keep this pull request like this. But let me add a Todo for myself, before this is ready to merge:

- [ ] Document how to use PHPStan in CONTRIBUTING.md

**Comment by szepeviktor** [comment]
> before we apply the WordPress Coding Standards

@Automatic uses Neutron. Give it a try!
https://github.com/Automattic/phpcs-neutron-ruleset

> Do you know why there are so many errors in Travis?

In PHPStan config we have `level: max`

**Comment by szepeviktor** [comment]
I work for developer teams. I talk them into using the following zero-coverage tests in CI:

1. syntax check
2. coding standard (much more than how you indent your code!)
3. static analysis

It makes your code robust without writing a single test.

**Comment by jarednova** [comment]
@gchtr it looks like something in #2146 broke some stuff. It's also possible it fixed something that was suppressing legit errors from before. Looking into it now!

**Comment by jarednova** [comment]
Found the root error and fixed in #2154. Once someone else confirms we can merge that in across 2.x which should cure the mass of Travis errors

**Comment by jarednova** [comment]
Thanks @szepeviktor ! Got this into Travis, let's see how it goes. 

What I'm imagining is that in this PR we can repair the few errors that appear at level `0`. Once merged, we can work our way up levels as you suggest

**Comment by mahagr** [comment]
It is a huge task to fix all the phpstan errors, but for the projects where I've done that, it was totally worth it.

Another great thing with phpstan is that it allows you to fix bad code documentation, thus making it easier to use the library from IDEs.

Just be careful with it, as fixing the issues will break some badly written (but kinda legit) code.

**Comment by szepeviktor** [comment]
Level 0: https://travis-ci.org/timber/timber/jobs/634133135#L3463
Level 2: I think it takes 30 minutes.

**Comment by szepeviktor** [comment]
@jarednova I've been [told not to fix errors](https://github.com/timber/timber/pull/1986#issuecomment-571307148).

Do we need CI to pass in this PR?

**Comment by jarednova** [comment]
@szepeviktor oh crap, you're right — we did say that! Here's what I'm thinking on this then ...

- For now, I disabled Travis _running_ the static analysis tests (but all the code to do so is in there and seems to work as intended)
- In this PR: Document how to use PHPStan in CONTRIBUTING.md (I'm happy to take a first pass at this)
- Let's merge this in and in a separate PR I'll pull in a7da8fd
- After that's merged, consider whether to submit other PRs as we keep raising up to levels 1, 2, 3. BUT it might make sense to wait on these until 2.x is stable, has factories merged in, etc.

I think my concern (from being burned in the past on stuff like this) is a big hairy set of merge conflicts that are going to potentially corrupt a ton of work or just make life difficult as we try to get stuff set. Before going with this, want to make sure to get @gchtr's take.

Thanks for your work on this @szepeviktor. As @mahagr suggests, I'm hopeful that it will make everything run smoother/faster as we move ahead!

**Comment by szepeviktor** [comment]
@jarednova Could you merge it before any nasty conflicts arise?

**Comment by gchtr** [comment]
> * In this PR: Document how to use PHPStan in CONTRIBUTING.md (I'm happy to take a first pass at this)

@jared I’d be happy if you could do this!

> * Let's merge this in and in a separate PR I'll pull in [a7da8fd](https://github.com/timber/timber/commit/a7da8fd38b8b970ad94c415daf384ed4d2241b43)
> * After that's merged, consider whether to submit other PRs as we keep raising up to levels 1, 2, 3. BUT it might make sense to wait on these until 2.x is stable, has factories merged in, etc.

That sounds like a reasonable approach! Let’s wait for the big bunch of factory pull requests to be merged in before we start working on PHPStan fixes 👍.

**Comment by szepeviktor** [comment]
Okay.

File: docs/guides/testing.md [hunk 1]
```diff
@@ -109,3 +109,18 @@ sudo /bin/dd if=/dev/zero of=/var/swap.1 bs=1M count=1024
...
+
+Version 2.x is also covered by static analysis courtesy of [PHPStan](https://github.com/phpstan/phpstan). This analyzes our code files and structure to ensure things like existence and accessibility of called methods and functions, number of passed arguments, data types in docs, etc.
+
+To run, use:
+
+```
+vendor/bin/phpstan analyze -l 1
+```
+
+Where the last argument is the level of strictness (0-4) to apply. We're currently starting at level 1 and working our way up over time.
```
Comment (gchtr) [inline docs/guides/testing.md]: I think there are 8 levels, right? Or does this mean that we only intend to reach level 4 for the moment (which I think would also be reasonable).

**Comment by jarednova** [comment]
@szepeviktor I wrote up a quick guide for local usage and how we're implementing. Can you give it a quick 👍 / 👎 (or edit) to make sure I describe that tool correctly? https://github.com/timber/timber/pull/2149/files#diff-2c4188d6f30d75cb594a9522dc2add4fR112-R126

File: docs/guides/testing.md (hunk 1 above)
Comment (jarednova) [reply docs/guides/testing.md]: good call, I copied that from the initial release posting and not the newest levels of support

Comment (gchtr) [reply docs/guides/testing.md]: Cool, so 7 it is, good to know 😄

**Pull request merged**